### PR TITLE
Add boundary conditions to Block and Domain classes

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Have `Block` hold a `DirectionMap<std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>>`. Currrently cannot enforce that these are set correctly because no domain creators support boundary conditions.
- Add support for boundary conditions to the `Domain` class and various helper functions that construct a `Domain`

The next step for supporting boundary conditions in the domain is adding support to the domain creators. At that point I think the elliptic and evolution codes can independently add support boundary conditions that are read from the input file.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Domain/Block.cpp [hunk 1]
```diff
@@ -19,15 +21,29 @@ Block<VolumeDim>::Block(
...
   for (const auto& direction : Direction<VolumeDim>::all_directions()) {
     if (neighbors_.find(direction) == neighbors_.end()) {
-      external_boundaries_.emplace(std::move(direction));
+      if (not external_boundary_conditions.contains(direction)) {
+        // Fill with a nullptr for now so we don't break existing code.
+        external_boundary_conditions_[direction] = nullptr;
+        // We cannot enable this error message until boundary conditions are
+        // enabled everywhere in the code, unfortunately.
+        // ERROR("Couldn't find boundary condition for external boundary "
+        //       << direction);
```
Comment (nilsvu) [inline src/Domain/Block.cpp:62]: I think you can already test the reverse case though: Make sure the user didn't pass a boundary condition for a direction that's actually internal

File: src/Domain/Domain.cpp [hunk 2]
```diff
@@ -27,13 +27,28 @@ template <size_t VolumeDim>
...
         domain::CoordinateMapBase<Frame::Logical, Frame::Inertial, VolumeDim>>>
-        maps) noexcept {
+        maps,
+    std::vector<DirectionMap<
+        VolumeDim,
+        std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>>>
+        boundary_conditions) noexcept {
+  ASSERT(
+      boundary_conditions.size() == maps.size() or boundary_conditions.empty(),
+      "There must be either one boundary condition per block or none at all "
```
Comment (nilsvu) [inline src/Domain/Domain.cpp]: `...one set of boundary conditions per...` or similar (another occurrence below)

File: src/Domain/Block.hpp [hunk 3]
```diff
@@ -46,10 +47,15 @@ class Block {
...
+  ///        at external boundaries.
   Block(std::unique_ptr<domain::CoordinateMapBase<
             Frame::Logical, Frame::Inertial, VolumeDim>>&& stationary_map,
-        size_t id,
-        DirectionMap<VolumeDim, BlockNeighbor<VolumeDim>> neighbors) noexcept;
+        size_t id, DirectionMap<VolumeDim, BlockNeighbor<VolumeDim>> neighbors,
+        DirectionMap<
+            VolumeDim,
+            std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>>
+            external_boundary_conditions = {}) noexcept;
```
Comment (nilsvu) [inline src/Domain/Block.hpp:59]: Do you need this to have a default for a particular reason? Once the consistency check in the implementation can be enabled I can't think of a case where one would not have to think about the set of boundary conditions when constructing a Block.

File: src/Domain/Domain.hpp [hunk 4]
```diff
@@ -53,9 +53,14 @@ class Domain {
...
-                      Frame::Logical, Frame::Inertial, VolumeDim>>>
-                      maps) noexcept;
+  explicit Domain(
+      std::vector<std::unique_ptr<domain::CoordinateMapBase<
+          Frame::Logical, Frame::Inertial, VolumeDim>>>
+          maps,
+      std::vector<DirectionMap<
+          VolumeDim,
+          std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>>>
+          boundary_conditions = {}) noexcept;
```
Comment (nilsvu) [inline src/Domain/Domain.hpp:67]: same question about the necessity of this default. Also it would probably be good to add a note on the "specify all b.c. or none at all" requirement to the docs for this class.

File: src/Domain/DomainHelpers.cpp [hunk 5]
```diff
@@ -310,7 +310,7 @@ std::vector<std::array<size_t, two_to_the(VolumeDim)>> corners_from_two_maps(
         gsl::at(corners_for_block2, vci_map2.local_corner_number()) =
             vci_map1.local_corner_number();
         break;
-      } else {
+      } else {  // NOLINT(llvm-else-after-return)
```
Comment (nilsvu) [inline src/Domain/DomainHelpers.cpp]: llvm has a fair point here, don't you think? I understand if you don't want to refactor this code though. Instead of the NOLINT we can just merge and ignore the clang-tidy complaint.

File: tests/Unit/Domain/Test_Domain.cpp [hunk 6]
```diff
@@ -235,54 +271,86 @@ void test_1d_domains() {
...
+template <size_t Dim>
+auto compute_boundary_conditions(
+    const std::vector<std::unordered_set<Direction<Dim>>>&
+        external_boundaries) noexcept {
+  using MapType = DirectionMap<
+      Dim, std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>>;
+  std::vector<MapType> boundary_conditions{external_boundaries.size()};
+  for (size_t i = 0; i < external_boundaries.size(); ++i) {
+    const auto& external_boundaries_of_block = external_boundaries[i];
+    MapType t{};
```
Comment (nilsvu) [inline tests/Unit/Domain/Test_Domain.cpp]: t -> boundary_conditions_i ?

**Review by nilsvu — CHANGES_REQUESTED** [review]
Two design questions that we can also discuss in a call if you prefer:

1. I missed to ask in #2750: Doesn't the `domain::BoundaryConditions::BoundaryCondition` base class need the list of registrars as template parameter? If I understand correctly it needs to build the `creatable_classes` list so it can be factory-created from domain creators.
2. Do you have an idea how to handle cases where boundary conditions are not needed, e.g. for things like `ExportCoordinates` or when we create an `ApplyObservers`-type executable or do interpolations between domains?

File: src/Domain/Block.hpp (hunk 3 above)
Comment (nilsdeppe) [reply src/Domain/Block.hpp:59]: Yes, once everything is updated we don't need the default, but I wasn't really feeling motivated to find all the defaults.

File: src/Domain/DomainHelpers.cpp (hunk 5 above)
Comment (nilsdeppe) [reply src/Domain/DomainHelpers.cpp]: I don't understand what LLVM is trying to tell me. As far as I can tell it wants me to remove the `break;` but that actually changes the control flow.

**Comment by nilsdeppe** [comment]
- This doesn't work without templating the domain on the system or metavariables, which I don't think we should do. Instead, each system defines another base class that is used for the factor. The interval domain creator will have tags:

  ```cpp
  template <typename BoundaryConditionsBase>
  struct UpperBoundaryCondition {
    static std::string name() noexcept { return "UpperBoundary"; }
    static constexpr Options::String help =
        "Options for the boundary condition applied at the upper boundary.";
    using type = std::unique_ptr<BoundaryConditionsBase>;
    using group = BoundaryConditions;
  };
  template <typename BoundaryConditionsBase>
  struct LowerBoundaryCondition {
    static std::string name() noexcept { return "LowerBoundary"; }
    static constexpr Options::String help =
        "Options for the boundary condition applied at the lower boundary.";
    using type = std::unique_ptr<BoundaryConditionsBase>;
    using group = BoundaryConditions;
  };

  using basic_options =
      tmpl::list<LowerBound, UpperBound, IsPeriodicIn, InitialRefinement,
                 InitialGridPoints, TimeDependence>;

  template <typename Metavariables>
  using options = tmpl::conditional_t<
      domain::BoundaryConditions::has_boundary_conditions_base_v<
          typename Metavariables::system>,
      tmpl::push_back<
          basic_options,
          LowerBoundaryCondition<
              domain::BoundaryConditions::get_boundary_conditions_base<
                  typename Metavariables::system>>,
          UpperBoundaryCondition<
              domain::BoundaryConditions::get_boundary_conditions_base<
                  typename Metavariables::system>>>,
      basic_options>;
  ```

  Basically, the system has a `using boundary_conditions_base =...` type alias that points to the base class that is factory-createable. This base class much inherit from `domain::BoundaryConditions::BoundaryCondition`
- They will need to set up some empty system with an a `None` boundary condition. We could make specifying boundary conditions optional, but that seems fragile. As per `ApplyObservers`: I think that should probably have access to the boundary conditions. One may conceivably want to observe the boundary condition terms being added to the time derivatives.

File: src/Domain/DomainHelpers.cpp (hunk 5 above)
Comment (nilsvu) [reply src/Domain/DomainHelpers.cpp]: I think it wants you to delete the `else` scope, moving the code one indentation level down. The `else` doesn't really do anything here because the `break` in the `if` means the code below doesn't run anyway when the `if` is triggered.

Comment (nilsdeppe) [reply src/Domain/DomainHelpers.cpp]: Ah, I'll try that and see if it works.

**Comment by nilsvu** [comment]
I see, thanks for explaining. I agree we shouldn't template the domain on the metavariables or the system, that seems bad. We _could_ template the domain creators on the boundary condition base class when we update them to support boundary conditions. I'm suggesting this because I don't like the pattern of "reaching up" to the metavariables to retrieve a special type alias and much prefer the pattern of "passing down" such information from the metavariables. If the "reaching up" pattern seems unavoidable then I would suggest to make the type alias part of the metavariables, not the system. The reason is that multiple executables may use the same system but want to specify a different set of possible boundary conditions. For example, a `SolveBBH` executable offers the choice of apparent-horizon boundary conditions whereas a `SolveBNS` executable does not. Similarly, a `SolveThermalNoise` executable offers a `LaserBeam` boundary condition but a `SolveBentMaterial` executable does not, but both solve an elasticity system.

**Comment by nilsdeppe** [comment]
What I don't like about templating all of the domain creator on the boundary conditions is that this means either more code in hpp files (bad for compile time) or _a lot_ of `tpp` files that require many explicit instantiations (not super user friendly, I think). I think the benefit of adding the boundary conditions as a template parameter to the domain creators is one could easily solve different systems on different domains. This _might_ be of interest when we solve the Boltzmann equation, but I think there are way too many other problems to figure out before then to make that a compelling argument for making the code harder to use now.

I intentionally didn't add the type alias to the `metavariables` since boundary conditions are valid for specific systems only. I also find having lots of different options listed as separate type aliases in the metavariables makes the code impossible to understand. From your list: `SolveBBH` and `SolveBNS` are solving different physical systems, one has matter one does not. You could template the `System` class on boundary conditions if you want to reduce code, but I think they should not be treated as the same system. The second example: you are trying to restrict the boundary conditions for the user. While reasonable, I don't think we should make the evolution code harder to use because of this. Further more, if I can specify the domain, initial conditions, and boundary conditions in the input file, doesn't that mean you can combine both the executables you listed into a single executable? This is what the evolution code is going to do, and then it is up to the user to specify a reasonable physical setup.

In short, I like the current design, if you want to restrict what boundary conditions are available I think you should template the system on boundary conditions.

**Comment by nilsvu** [comment]
+1 to keeping domain creator code in cpp files. Boundary conditions should disrupt the domain code as little as possible. The particular base class is only needed for the option-creation process, and I also don't see another way of injecting compile-time information into the option-creation without going though the metavariables.

About metavars vs system: We might end up with exactly one executable per system, but I don't think we know that for sure right now. Seems too early to make that call..

**Comment by nilsdeppe** [comment]
Yes, but at least in the evolution there are many systems (e.g. Newtonian Euler) where the evolution larger differs only by the domain, boundary conditions, and initial condition. We will likely have separate executables for BBH and single BH, but that's more so because of the AH finders and control systems rather than the system/domain. Does templating the system on the boundary conditions work for your use case?

**Comment by nilsvu** [comment]
Sure it works to template the system on the boundary conditions, but then all it does is receive the template parameter and expose it as an alias, which seems silly 😄 Perhaps I could make the system use all boundary conditions by default and allow a template parameter to restrict the set.

**Comment by nilsdeppe** [comment]
Okay :)

File: src/Domain/DomainHelpers.cpp (hunk 5 above)
Comment (wthrowe) [reply src/Domain/DomainHelpers.cpp]: Looks like a rename of readability-else-after-return, which is in our exclusion list.

**Comment by nilsdeppe** [comment]
pushed fixups and a new commit that adds the `llvm-...` to the clang-tidy ignore list

File: src/Domain/Block.cpp [hunk 7]
```diff
@@ -28,6 +28,11 @@ Block<VolumeDim>::Block(
     : stationary_map_(std::move(stationary_map)),
       id_(id),
       neighbors_(std::move(neighbors)) {
+  ASSERT(external_boundary_conditions.empty() or
+             external_boundary_conditions.size() ==
+                 2 * VolumeDim - neighbors_.size(),
```
Comment (nilsvu) [inline src/Domain/Block.cpp:41]: This assert doesn't catch the case where `external_boundary_conditions` has the correct number of entries, but at the wrong directions. I suggest to test that, perhaps wrap the loop in `#ifdef SPECTRE_DEBUG` if necessary

File: src/Domain/DomainHelpers.cpp [hunk 8]
```diff
@@ -1303,7 +1318,8 @@ ShellWedges Options::create_from_yaml<ShellWedges>::create<void>(
   const auto which_wedges = options.parse_as<std::string>();
   if (which_wedges == "All") {
     return ShellWedges::All;
-  } else if (which_wedges == "FourOnEquator") {
+  } else if (which_wedges ==  // NOLINT(llvm-else-after-return)
```
Comment (nilsvu) [inline src/Domain/DomainHelpers.cpp]: here's a left-over NOLINT

**Review by nilsvu — CHANGES_REQUESTED** [review]
Please squash, including these

**Comment by nilsdeppe** [comment]
Rebased and squashed. I added an `ASSERT` inside the existing for loop that checks that either there are no boundary conditions at all or the current external direction has a boundary condition specified. Thanks for the review :)

File: src/Domain/Block.cpp [hunk 9]
```diff
@@ -98,8 +125,21 @@ std::ostream& operator<<(std::ostream& os,
 template <size_t VolumeDim>
 bool operator==(const Block<VolumeDim>& lhs,
                 const Block<VolumeDim>& rhs) noexcept {
+  // Since the external boundary conditions are all abstract base classes we
+  // can't compare them. We at least ensure that they match in terms of the
+  // number of external boundaries set.
   return lhs.id() == rhs.id() and lhs.neighbors() == rhs.neighbors() and
          lhs.external_boundaries() == rhs.external_boundaries() and
+         alg::all_of(rhs.external_boundaries(),
```
Comment (kidder) [inline src/Domain/Block.cpp]: I don't see how this tests anything about the equivalence of Blocks since `lhs.external_boundaries() == rhs.external_boundaries()`.  It will only fail for an invalid Block which should be caught at construction...

Comment (nilsdeppe) [reply src/Domain/Block.cpp]: heh, yea I think you're right. The other test I can add is that the `typeid` of the dereferenced base class pointers are the same to make sure that at the least the boundary conditions are of the same type. Thoughts?

Comment (kidder) [reply src/Domain/Block.cpp]: seems reasonable

**Comment by nilsdeppe** [comment]
Rebased and squashed in the additional comparisons

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
@nilsleiffischer please look at

File: src/Domain/Block.cpp [hunk 10]
```diff
@@ -98,8 +125,19 @@ std::ostream& operator<<(std::ostream& os,
...
+  // RHS.
   return lhs.id() == rhs.id() and lhs.neighbors() == rhs.neighbors() and
          lhs.external_boundaries() == rhs.external_boundaries() and
+         alg::all_of(
+             rhs.external_boundaries(),
+             [&lhs, &rhs](const Direction<VolumeDim>& dir) noexcept {
+               return lhs.external_boundary_conditions().contains(dir) and
+                      rhs.external_boundary_conditions().contains(dir) and
+                      typeid(lhs.external_boundary_conditions().at(dir)) ==
+                          typeid(rhs.external_boundary_conditions().at(dir));
```
Comment (nilsvu) [inline src/Domain/Block.cpp:147]: I believe you have to `#include <typeinfo>` to use `typeid`. While you're at it you might as well add the includes for `Direction`, `DirectionMap`, `<memory>`, `BlockNeighbor`, `<utility>`

File: tests/Unit/Domain/Test_Block.cpp [hunk 11]
```diff
@@ -26,19 +27,72 @@
 #include "Domain/Structure/DirectionMap.hpp"
 #include "Domain/Structure/OrientationMap.hpp"
 #include "Framework/TestHelpers.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Utilities/CloneUniquePtrs.hpp"
```
Comment (nilsvu) [inline tests/Unit/Domain/Test_Block.cpp]: I think you don't actually use `clone_unique_ptrs`

File: src/Domain/Domain.hpp [hunk 12]
```diff
@@ -52,10 +52,17 @@ class Domain {
...
+   * boundary conditions at all.
    */
-  explicit Domain(std::vector<std::unique_ptr<domain::CoordinateMapBase<
-                      Frame::Logical, Frame::Inertial, VolumeDim>>>
-                      maps) noexcept;
+  explicit Domain(
+      std::vector<std::unique_ptr<domain::CoordinateMapBase<
+          Frame::Logical, Frame::Inertial, VolumeDim>>>
+          maps,
+      std::vector<DirectionMap<
```
Comment (nilsvu) [inline src/Domain/Domain.hpp:64]: While squashing, add include for `DirectionMap` and `domain::BoundaryConditions::BoundaryCondition`.

File: src/Domain/Block.cpp [hunk 13]
```diff
@@ -19,15 +21,39 @@ Block<VolumeDim>::Block(
...
   // set these Directions to external_boundaries_.
   for (const auto& direction : Direction<VolumeDim>::all_directions()) {
     if (neighbors_.find(direction) == neighbors_.end()) {
-      external_boundaries_.emplace(std::move(direction));
+      ASSERT(external_boundary_conditions.empty() or
+                 external_boundary_conditions.contains(direction),
+             "Specifying boundary conditions but the external direction "
+                 << direction << " in block " << id_
+                 << " does not have a specified boundary condition.");
+      if (not external_boundary_conditions.contains(direction)) {
```
Comment (nilsvu) [inline src/Domain/Block.cpp]: (optional) I find it easier to follow control flow with fewer boolean inversions, so I suggest to swap the `if` and `else` clauses here.

**Review by nilsvu — COMMENTED** [review]
Please squash these in directly

**Comment by nilsdeppe** [comment]
rebased and squashed in changes. I removed `<ostream>` from `Block.hpp`, replacing it with `<iosfwd>` since that's all it needs. Thanks for the reviews!

**Review by nilsvu — APPROVED** [review]

**Review by kidder — APPROVED** [review]
